### PR TITLE
render: remove wlr_renderer_check_import_dmabuf

### PR DESCRIPTION
**Pull request by emersion**

### CODE REVIEW

File: types/wlr_linux_dmabuf.c [hunk 1]
```diff
@@ -212,7 +225,7 @@ static void params_create_common(struct wl_client *client,
 
 		// planes > 0 might be subsampled according to fourcc format
 		if (i == 0 && buffer->attributes.offset[i] +
-				buffer->attributes.stride[i] * height >= size) {
+				buffer->attributes.stride[i] * height > size) {
```
Comment (agx) [inline types/wlr_linux_dmabuf.c]: Isn't >= correct here since we're starting at 0? In any case this should be in a separate commit.

Comment (emersion) [reply types/wlr_linux_dmabuf.c]: Sooo. I changed this in my previous PR (#1015) and it breaks `weston-simple-dmabuf-drm --import-format=XRGB` (from weston HEAD). Do you think it's a weston bug?

Comment (agx) [reply types/wlr_linux_dmabuf.c]: Uhh...I missed that change in your previous PR. 
The current code ```>=``` is more strict than just ```>``` or am I missing something? So I wonder how using ```>=``` would make something work that didn't work with just ```>``` ?

Anyways, I'd drop that change.

Comment (emersion) [reply types/wlr_linux_dmabuf.c]: >The current code >= is more strict than just > or am I missing something?

Yeah, the current code is more strict because it makes sense and should work - see all previous checks.

>So I wonder how using >= would make something work that didn't work with just > ?

The current code (`>=`) breaks `weston-simple-dmabuf-drm`. So this change (`>`) unbreaks it.

>Anyways, I'd drop that change.

By "drop", do you mean keep it as strict or make it more lenient just like before #1015?

Comment (agx) [reply types/wlr_linux_dmabuf.c]: Rereading the code the ```>=``` version is correct. I missed the ```* height``` there. So let's just move the part that turns ```>``` into ```>=``` into a separate commit so we don't sneak in any unrelated changes.

Comment (agx) [reply types/wlr_linux_dmabuf.c]: I flipped ```>=``` with ```>``` in the above (sorry, to many things going on). I meant to say: let's make it ```>``` in a separate commit since that's correct. stride * height == size is o.k.

Comment (emersion) [reply types/wlr_linux_dmabuf.c]: Okay. Is there any explanation why `stride * height == size` is OK? I assume we'd need to update the check above this one as well, in case of images with height=1.

Comment (agx) [reply types/wlr_linux_dmabuf.c]: Take the extreme case of height == 1 and width == stride == size. We have pixels [0, width-1] but ```offset + stride == size```. (so maybe the check should read

```
offset + (stride * height) -1 >= size
```

which is

```
offset + (stride * height) > size
```

Comment (emersion) [reply types/wlr_linux_dmabuf.c]: Oh yeah, that does make sense

**Comment by emersion** [comment]
Fixed bounds checking. Test plan: `weston-simple-dmabuf-drm --import-format={XRGB,NV12}` plus try some Xwayland apps.

File: types/wlr_linux_dmabuf.c [hunk 2]
```diff
@@ -207,15 +207,16 @@ static void params_create_common(struct wl_client *client,
 			// Skip checks if kernel does no support seek on buffer
 			continue;
 		}
-		if (buffer->attributes.offset[i] >= size) {
+		if (buffer->attributes.offset[i] > size) {
```
Comment (agx) [inline types/wlr_linux_dmabuf.c]: The ```>=``` is here to make sure we have at least some data to read. Changing it to ```>``` means we're o.k. with having an offset but nothing left to read. I'd prefere to leave it as is but I'm o.k. either way.

File: types/wlr_linux_dmabuf.c [hunk 3]
```diff
@@ -207,15 +207,16 @@ static void params_create_common(struct wl_client *client,
...
 			wl_resource_post_error(params_resource,
 				ZWP_LINUX_BUFFER_PARAMS_V1_ERROR_OUT_OF_BOUNDS,
 				"invalid offset %i for plane %d",
 				buffer->attributes.offset[i], i);
 			goto err_out;
 		}
 
-		if (buffer->attributes.offset[i] + buffer->attributes.stride[i]	> size) {
+		if (buffer->attributes.offset[i] + buffer->attributes.stride[i] > size ||
+				buffer->attributes.stride[i] == 0) {
```
Comment (agx) [inline types/wlr_linux_dmabuf.c]: That makes sense, also I wonder if we should reject stride < width?

File: types/wlr_linux_dmabuf.c (hunk 1 above)
Comment (agx) [reply types/wlr_linux_dmabuf.c]: Looks good now. Thanks!

File: types/wlr_linux_dmabuf.c (hunk 2 above)
Comment (emersion) [reply types/wlr_linux_dmabuf.c]: Hmm, the intention was to check here that the offset is okay, and in the next condition that the it's not empty.

File: types/wlr_linux_dmabuf.c (hunk 3 above)
Comment (emersion) [reply types/wlr_linux_dmabuf.c]: I'd rather not do that because there are obscure formats which have less than one byte per pixel.

Comment (agx) [reply types/wlr_linux_dmabuf.c]: Yeah, it would only work for plane 0

**Comment by emersion** [comment]
Thanks for your review!
